Docs update (#73)
solidity-labs-io/forge-proposal-simulator: docs/mainnet-examples/OptimismMultisig.md

-   `description()`: Provide a detailed description of your proposal.
-   `deploy()`: This example demonstrates the deployment of new MockUpgrade which will be
    used as new implementation to proxy.
-   `run()`: Sets environment for running the proposal. It sets `addresses`. `addresses` is address object
    containing addresses to be used in proposal that are fetched from `Addresses.json`.
-   `build()`: In this example, L1 NFT Bridge is upgraded to new implementation. The actions
    should be written in solidity code and in the order they should be executed. Any calls (except to the Addresses object) will be recorded and stored as actions to execute in the run function. `caller` address is passed into `buildModifier` that will call actions in `build`. Caller is optimism multisig for this example.
-   `simulate()`: Execute the proposal actions outlined in the `build()` step. This
    function performs a call to `_simulateActions()` from the inherited
    `MultisigProposal` contract. Internally, `_simulateActions()` simulates a call to the [Multicall3](https://www.multicall3.com/) contract with the calldata generated from the actions set up in the build step.

> -   `validate()`: It validates implementation is upgraded correctly.

Reviewer: There are a lot of places where these steps are essentially repeated, so these should be consolidated in places where they can for brevity. 

Seems like there are a lot of overlap between `guides` and `examples`

This content should be differentiated, and we shouldn't be repeating ourselves so much.

There should be a single place where each step of the proposal is explained, and then the individual docs talk about how this proposal's version of that function implementation differs or works.

For the examples, we should show Solidity code snippets at each step to demonstrate to developers how this works under the hood, and draw out the state machines so they can see how each step does or does not mutate state. All assumptions should be explicitly stated, nothing should be left to chance.

Author: > There are a lot of places where these steps are essentially repeated, so these should be consolidated in places where they can for brevity.

These are already consolidated here [proposal-functions.md](https://github.com/solidity-labs-io/forge-proposal-simulator/blob/feat/docs-update/docs/overview/architecture/proposal-functions.md)

> This content should be differentiated, and we shouldn't be repeating ourselves so much. There should be a single place where each step of the proposal is explained, and then the individual docs talk about how this proposal's version of that function implementation differs or works.

I think this repetition is needed because a user might go through only a single example and it will be easier for that user to walk through that example if we repeat these details.

> For the examples, we should show Solidity code snippets at each step to demonstrate to developers how this works under the hood, and draw out the state machines so they can see how each step does or does not mutate state. All assumptions should be explicitly stated, nothing should be left to chance.

Agree with this. Making this update.

Reviewer: > I think this repetition is needed because a user might go through only a single example and it will be easier for that user to walk through that example if we repeat these details.

That makes sense

## Running the Proposal

```sh
forge script mocks/MockMultisigProposal.sol --fork-url mainnet
```

> It's crucial to ensure all required address is correctly listed in the Addresses.json file. If these don't align, the script execution will fail.

Reviewer: wording, already commented on this above

Author: done

> # Optimism multisig Proposal

Reviewer 2: let's also add Arbitrum example using GovernorOZProposal from fps-example-repo

# Optimism Multisig Proposal

## Overview

> This is an example of FPS where FPS is used to make proposals for the Optimism Multisig on mainnet. This example upgrades the L1 NFT Bridge contract. The Optimism Multisig calls `upgrade` on the proxy contract to upgrade the implementation to a new `MockUpgrade`.

Reviewer: ```suggestion
This is an example where FPS is used to make proposals for the Optimism Multisig on mainnet. This example upgrades the L1 NFT Bridge contract. The Optimism Multisig calls `upgrade` on the proxy contract to upgrade the implementation to a new `MockUpgrade`.
```

-   `name()`: Defines the name of your proposal.

    ```solidity
    function name() public pure override returns (string memory) {
        return "OPTIMISM_MULTISIG_MOCK";
    }
    ```

-   `description()`: Provides a detailed description of your proposal.

    ```solidity
    function description() public pure override returns (string memory) {
        return "Mock proposal that upgrades the L1 NFT Bridge";
    }
    ```

-   `deploy()`: This example demonstrates the deployment of the new MockUpgrade, which will be used as the new implementation for the proxy.

    ```solidity
    function deploy() public override {
        if (!addresses.isAddressSet("OPTIMISM_L1_NFT_BRIDGE_IMPLEMENTATION")) {
            address l1NFTBridgeImplementation = address(new MockUpgrade());

            addresses.addAddress(
                "OPTIMISM_L1_NFT_BRIDGE_IMPLEMENTATION",
                l1NFTBridgeImplementation,
                true
            );
        }
    }
    ```

    Since these changes do not persist from runs themselves, after the contracts are deployed, the user must update the Addresses.json file with the newly deployed contract addresses.

> -   `build()`: Add actions to the proposal contract. [See the build function](../overview/architecture/proposal-functions.md#build-function). In this example, the L1 NFT Bridge is upgraded to a new implementation. The actions should be written in solidity code and in the order they should be executed. Any calls (except to the Addresses object) will be recorded and stored as actions to execute in the run function. The `caller` address is passed into `buildModifier` that will call actions in `build`. The caller is the Optimism Multisig for this example. The `buildModifier` is a necessary modifier for the `build` function and will not work without it.

Reviewer: same comments as preious prop

> -   `build()`: Add actions to the proposal contract. [See build function](../overview/architecture/proposal-functions.md#build-function). In this example, the L1 NFT Bridge is upgraded to a new implementation. The actions should be written in solidity code and in the order they should be executed. Any calls (except to the Addresses object) will be recorded and stored as actions to execute in the run function. The The `caller` address is passed into `buildModifier` that will call actions in `build`. The caller is the Optimism Multisig for this example. `buildModifier` is a necessary modifier for the `build` function and will not work without it.

Reviewer: ```suggestion
-   `build()`: Add actions to the proposal contract. [See build function](../overview/architecture/proposal-functions.md#build-function). In this example, the L1 NFT Bridge is upgraded to a new implementation. The actions should be written in Solidity code and in the order they should be executed. Any calls (except to the Addresses or Foundry Vm contract) will be recorded and stored as actions to execute in the run function. The The `caller` address is passed into `buildModifier` that will call actions in `build`. The caller is the Optimism Multisig for this example. `buildModifier` is a necessary modifier for the `build` function and will not work without it.
```

Reviewer: remove double the

    ```solidity
    function build()
        public
        override
        buildModifier(addresses.getAddress("OPTIMISM_MULTISIG"))
    {
        /// STATICCALL -- not recorded for the run stage
        IProxyAdmin proxy = IProxyAdmin(
            addresses.getAddress("OPTIMISM_PROXY_ADMIN")
        );

        /// CALLS -- mutative and recorded
        proxy.upgrade(
            addresses.getAddress("OPTIMISM_L1_NFT_BRIDGE_PROXY"),
            addresses.getAddress("OPTIMISM_L1_NFT_BRIDGE_IMPLEMENTATION")
        );
    }
    ```

> -   `run()`: Sets up the environment for running the proposal. [See run function](../overview/architecture/proposal-functions.md#run-function). This sets `addresses`, `primaryForkId`, and calls `super.run()` to run the entire proposal. In this example, `primaryForkId` is set to `mainnet` and selecting the fork for running the proposal. Next, the `addresses` object is set by reading `addresses.json` file.

Reviewer: ```suggestion
-   `run()`: Sets up the environment for running the proposal. [See run function](../overview/architecture/proposal-functions.md#run-function). This sets `addresses`, `primaryForkId`, and calls `super.run()` to run the entire proposal. In this example, `primaryForkId` is set to `mainnet` and selecting the fork for running the proposal. Next, the `addresses` object is set by reading from the `Addresses.json` file.
```

> -   `run()`: Sets up the environment for running the proposal. [See the run function](../overview/architecture/proposal-functions.md#run-function). This sets `addresses`, `primaryForkId`, and calls `super.run()` to run the entire proposal. In this example, `primaryForkId` is set to `mainnet` and selecting the fork for running the proposal. Next, the `addresses` object is set by reading from the `Addresses.json` file.

Reviewer: same comments as preious prop

    ```solidity
    function run() public override {
        // Create and select mainnet fork for proposal execution.
        primaryForkId = vm.createFork("mainnet");
        vm.selectFork(primaryForkId);

        // Set addresses object reading addresses from JSON file.
        addresses = new Addresses(
            vm.envOr("ADDRESSES_PATH", string("./addresses/Addresses.json"))
        );

        // Call the run function of parent contract 'Proposal.sol'.
        super.run();
    }
    ```

-   `simulate()`: Executes the proposal actions outlined in the `build()` step. This function performs a call to `_simulateActions()` from the inherited `MultisigProposal` contract. Internally, `_simulateActions()` simulates a call to the [Multicall3](https://www.multicall3.com/) contract with the calldata generated from the actions set up in the build step.

    ```solidity

>     function simulate() public override {

Reviewer: what is the difference between simulate and run?

Author: `run()` is the entry point that runs all other functions including `simulate()`. `simulate()` simulates proposal actions that were added in the `build()` step. @ElliotFriedman

        // get multisig address
        address multisig = addresses.getAddress("OPTIMISM_MULTISIG");

        // simulate all actions in 'build' functions through multisig
        _simulateActions(multisig);
    }
    ```

-   `validate()`: Validates that the implementation is upgraded correctly.

    ```solidity
    function validate() public override {
        // get proxy address
        IProxy proxy = IProxy(
            addresses.getAddress("OPTIMISM_L1_NFT_BRIDGE_PROXY")
        );

        // implementation() caller must be the owner
        vm.startPrank(addresses.getAddress("OPTIMISM_PROXY_ADMIN"));

        // ensure implementation is upgraded
        require(
            proxy.implementation() ==
                addresses.getAddress("OPTIMISM_L1_NFT_BRIDGE_IMPLEMENTATION"),
            "Proxy implementation not set"
        );
        vm.stopPrank();
    }
    ```

## Running the Proposal

```sh
forge script mocks/MockMultisigProposal.sol --fork-url mainnet
```

> All required addresses should be in the Addresses.json file, including the `DEPLOYER_EOA` address, which will deploy the new contracts. If these don't align, the script execution will fail.

Reviewer: same comments as previous prop

    // simulate all actions in 'build' functions through multisig
    _simulateActions(multisig);
}
```

-   `validate()`: Validates that the implementation is upgraded correctly.

```solidity
function validate() public override {
    // get proxy address
    IProxy proxy = IProxy(addresses.getAddress("OPTIMISM_L1_NFT_BRIDGE_PROXY"));

    // implementation() caller must be the owner
    vm.startPrank(addresses.getAddress("OPTIMISM_PROXY_ADMIN"));

    // ensure implementation is upgraded
    require(
        proxy.implementation() ==
            addresses.getAddress("OPTIMISM_L1_NFT_BRIDGE_IMPLEMENTATION"),
        "Proxy implementation not set"
    );
    vm.stopPrank();
}
```

-   `run()`: Sets up the environment for running the proposal. [Refer](../overview/architecture/proposal-functions.md#run-function) It sets `addresses`, `primaryForkId`, and calls `super.run()` to run the proposal lifecycle. In this function, `primaryForkId` is set to `mainnet` and selecting the fork for running the proposal. Next, the `addresses` object is set by reading `addresses.json` file. The `addresses` contract state is persisted across forks using `vm.makePersistent()`.

```solidity

> function run() public override {

Reviewer: run before validate and immediately after simulate

Reviewer 3: done
